### DISCUSSION TITLE
How to check if in editmode/admin area?

### DISCUSSION THREAD

**Discussion by gvanto** [pimcore/pimcore#13603] [Q&A / Support] [answered]
Have googled this but cannot seem to find it... I have a (php) Listener class in which I need to know if i'm currently in admin edit page for a data object (similar to 'editmode' var in twig)... how to do this check in php?

Inside a Controller the editmode is available:

`$this->editmode`

 but how to get this in a class that's not a controller?

**Comment by kingjia90** [comment, marked as answer]
~By injecting `EditmodeResolver $editmodeResolver` and using `$this->editmodeResolver->isEditmode()` should work~
Answer: `PimcoreContextAwareTrait` `PimcoreContextResolver`

**Reply by gvanto** [reply]
Hey @kingjia90 I literally just tried doing that:

\Pimcore\Log\Simple::log('app',
            ' DBG: isEditMode=' . json_encode($this->editModeResolver->isEditmode()));

But, even though I am in pimcore admin data object edit page, this returns false (?) Expect it to return true...

**Reply by kingjia90** [reply]
ah, edit mode is just for documents(?)

**Reply by kingjia90** [reply]
what about `PimcoreContextAwareTrait`
`if ($this->matchesPimcoreContext($request, PimcoreContextResolver::CONTEXT_ADMIN)) {` ?

**Reply by gvanto** [reply]
> ah, edit mode is just for documents(?)

No idea man, can't see why it would only be for documents (?!)

**Reply by kingjia90** [reply]
> > ah, edit mode is just for documents(?)
> 
> No idea man, can't see why it would only be for documents (?!)

Ah, double-checked and for "edit mode" exclusively refers to the part in the documents where the input widgets (editables) are displayed in admin interface. That's why it won't work for objects.

**Reply by gvanto** [reply]
Sorry to bombard with issues, but now getting this: 
Error: "Missing pimcore context resolver. Is the listener properly configured?"

My code:
<img width="874" alt="image" src="https://user-images.githubusercontent.com/8372918/202447126-bb8a56e9-010e-4974-b61e-ccf7ae93c5e3.png">

Someone else also having similar issue:
https://github.com/pimcore/pimcore/issues/12886

**Reply by mattamon** [reply]
Couldn't you just check if is an admin session?
https://pimcore.com/docs/pimcore/current/Development_Documentation/Development_Tools_and_Details/Working_with_Sessions.html -> last line.
```php
Pimcore\Tool\Session::getReadonly()
```

**Reply by kingjia90** [reply]
Probably this shouldn't go in the constructor, but `onKernelResponse` or request, due the `setPimcoreContextResolver` setter injection called after constructor

**Reply by gvanto** [reply]
> Couldn't you just check if is an admin session? https://pimcore.com/docs/pimcore/current/Development_Documentation/Development_Tools_and_Details/Working_with_Sessions.html -> last line.
> 
> ```
> Pimcore\Tool\Session::getReadonly()
> ```

This returns an empty object, so probably not way to go.

**Reply by mattamon** [reply]
It should have attributes though, if you're logged in.
I checked for the user like this and get the following result for the user
```php
 $user = $session->get('user');
 p_r($user);
```
```php
Pimcore\Model\User {#2008 ▼
  #dao: null
  #id: 2
  #parentId: 0
  #name: "admin"
  #type: "user"
  -activeDispatchingEvents: []
  #permissions: array:1 [▶]
  #workspacesAsset: []
  #workspacesObject: []
  #workspacesDocument: []
  #classes: []
  #docTypes: []
  #perspectives: []
  #websiteTranslationLanguagesView: []
  #websiteTranslationLanguagesEdit: []
  #password: "..."
  #firstname: null
  #lastname: null
  #email: null
  #language: "en"
  #admin: true
  #active: true
  #roles: []
  #welcomescreen: false
  #closeWarning: true
  #memorizeTabs: true
  #allowDirtyClose: false
  #contentLanguages: null
  #activePerspective: null
  #mergedPerspectives: null
  #mergedWebsiteTranslationLanguagesEdit: null
  #mergedWebsiteTranslationLanguagesView: null
  #lastLogin: 1668610912
  #keyBindings: null
  #twoFactorAuthentication: array:4 [▶]
  id: 2
  parentId: 0
  name: "admin"
  type: "user"
  permissions: array:1 [▶]
  workspacesAsset: []
  workspacesObject: []
  workspacesDocument: []
  classes: []
  docTypes: []
  perspectives: []
  websiteTranslationLanguagesView: []
  websiteTranslationLanguagesEdit: []
  password: "$2y$10$etEOtBJiU85sWsSuZYtKsuxkPK1Qpck.i05Y41bZ8W.JrvWzZvTvG"
  firstname: null
  lastname: null
  email: null
  language: "en"
  admin: true
  active: true
  roles: []
  welcomescreen: false
  closeWarning: true
  memorizeTabs: true
  allowDirtyClose: false
  contentLanguages: null
  activePerspective: null
  mergedPerspectives: null
  mergedWebsiteTranslationLanguagesEdit: null
  mergedWebsiteTranslationLanguagesView: null
  lastLogin: 1668610912
  keyBindings: null
  twoFactorAuthentication: array:4 [▶]
}
```

If the user is set, there is an active admin_session. Of course there is no guarantee, if the user is in the backend or frontend.

**Reply by gvanto** [reply]
Exactly, user can be logged in as admin and visit the frontend...

Seems like this PimcoreContextResolver is best approach, thanks 👍
